polaris-core/src/main/java/org/apache/polaris/core/persistence/transactional/AbstractTransactionalPersistence.java

```diff
@@ -194,6 +117,547 @@ protected abstract void deleteFromEntitiesActive(
    * @param callCtx call context
    * @param entity entity record to delete
    */
-  protected abstract void deleteFromEntitiesChangeTracking(
+  protected abstract void deleteFromEntitiesChangeTrackingInCurrentTxn(
       @Nonnull PolarisCallContext callCtx, @Nonnull PolarisEntityCore entity);
+
+  //
+  // Implementations of the one-shot atomic BasePersistence methods which explicitly run
+  // the in-transaction variants of methods in a new transaction.
+  //
+
+  /** {@inheritDoc} */
+  @Override
+  public long generateNewId(@Nonnull PolarisCallContext callCtx) {
+    return runInTransaction(callCtx, () -> this.generateNewIdInCurrentTxn(callCtx));
+  }
+
+  /** Helper to perform the compare-and-swap semantics of a single writeEntity call. */
+  private void checkConditionsForWriteEntityInCurrentTxn(
+      @Nonnull PolarisCallContext callCtx,
+      @Nonnull PolarisBaseEntity entity,
+      @Nullable PolarisBaseEntity originalEntity) {
+    PolarisBaseEntity refreshedEntity =
+        this.lookupEntityInCurrentTxn(
+            callCtx, entity.getCatalogId(), entity.getId(), entity.getTypeCode());
+
+    if (originalEntity == null) {
+      if (refreshedEntity != null) {
+        // If this is a "create", and we manage to look up an existing entity with already
+        // the same id, where ids are uniquely reserved when generated, it means it's a
+        // low-level retry possibly in the face of a transient connectivity failure to
+        // the backend database.
+        throw new EntityAlreadyExistsException(refreshedEntity);
+      } else {
+        // Successfully verified the entity doesn't already exist by-id, but for a "create"
+        // we must also check for name-collection now.
+        refreshedEntity =
+            this.lookupEntityByNameInCurrentTxn(
+                callCtx,
+                entity.getCatalogId(),
+                entity.getParentId(),
+                entity.getType().getCode(),
+                entity.getName());
+        if (refreshedEntity != null) {
+          // Name-collision conflict.
+          throw new EntityAlreadyExistsException(refreshedEntity);
+        }
+      }
+    } else {
+      // This is an "update".
+      if (refreshedEntity == null
+          || refreshedEntity.getEntityVersion() != originalEntity.getEntityVersion()
+          || refreshedEntity.getGrantRecordsVersion() != originalEntity.getGrantRecordsVersion()) {
```

Comment:
I believe this approach implies that the transaction isolation level is "serializable", which is fine from my POV, I just want to confirm my understanding.

Reply:
Correct